File: packages/agent-python/shieldrasp/agent/circuit_breaker.py
import time

class CircuitBreaker:
    def __init__(self):
        self.timings = {}
        self.disabled_hooks = set()
        self.THRESHOLD_MS = 0.5
        self.SAMPLES = 100

    def record(self, hook_id: str, duration_ms: float):
        if hook_id in self.disabled_hooks:
            return
            
        if hook_id not in self.timings:
            self.timings[hook_id] = []
            
        stats = self.timings[hook_id]
        stats.append(duration_ms)
        
        if len(stats) > self.SAMPLES:
            stats.pop(0)
            
        if len(stats) == self.SAMPLES:
            avg = sum(stats) / self.SAMPLES
            if avg > self.THRESHOLD_MS:
                self.disable_hook(hook_id, avg)

    def disable_hook(self, hook_id: str, avg_duration: float):
        self.disabled_hooks.add(hook_id)

    def is_disabled(self, hook_id: str) -> bool:
        return hook_id in self.disabled_hooks

global_circuit_breaker = CircuitBreaker()

File: packages/agent/src/agent/circuit-breaker.ts
export class CircuitBreaker {
  private timings: Map<string, number[]> = new Map();
  private disabledHooks: Set<string> = new Set();
  private readonly THRESHOLD_MS = 0.5;
  private readonly SAMPLES = 100;

  record(hookId: string, durationMs: number) {
    if (this.disabledHooks.has(hookId)) return;
    
    let stats = this.timings.get(hookId);
    if (!stats) {
      stats = [];
      this.timings.set(hookId, stats);
    }
    
    stats.push(durationMs);
    if (stats.length > this.SAMPLES) {
      stats.shift();
    }

    if (stats.length === this.SAMPLES) {
      const avg = stats.reduce((a, b) => a + b, 0) / this.SAMPLES;
      if (avg > this.THRESHOLD_MS) {
        this.disableHook(hookId, avg);
      }
    }
  }

  private disableHook(hookId: string, avgDuration: number) {
    this.disabledHooks.add(hookId);
    console.warn(`[ShieldRASP] Circuit Breaker: Auto-disabled hook '${hookId}' (avg lag: ${avgDuration.toFixed(2)}ms)`);
  }

  isHookDisabled(hookId: string): boolean {
    return this.disabledHooks.has(hookId);
  }
}

export const globalCircuitBreaker = new CircuitBreaker();